AVM: EC math (#4924)
algorand/go-algorand: data/transactions/logic/TEAL_opcodes.md

| 0 | BlkSeed | []byte |  |
| 1 | BlkTimestamp | uint64 |  |

## ec_add v

- Opcode: 0xe0 {uint8 curve}
- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
- for (curve points A and B) return the curve point A + B
- **Cost**:  BN254=0 BN254_G1=10 BN254_G2=10 BLS12_381=0 BLS12_381_G1=20 BLS12_381_G2=20
- Availability: v9

`EC` Curves:

| Index | Name | Notes |
| - | ------ | --------- |

> | 0 | BN254 | BN254 curve, sometimes called BN256 |

Reviewer: 0 & 3 are not supported for `ec_add`

Author: Right, and `ec_pairing_check` only supports 0 and 3.  I suppose I should change things up so that the ops that there is a table of the G1/G2s, and a separate table for `ec_pairing_check`, which will only have 0 and 3 rows from this table.

There's no conceivable operation that would want to take either `BN254` OR `BN254_G1`, right?

Author: I've changed this so that only allowable curves are actual G1 and G2 curves of each kind.  No opcode accepts a field like 'BN256` (without the G1 or G2).  The only opcode that _did_ accept such an argument was `ec_pairing_check`. Now, that opcode accepts either the G1 _or_ the G2 form, and expects the named curve's points to be first on the stack, and then the "other" curve's points to be last (top) on stack.

| 1 | BN254_G1 | G1 of the BN254 curve |
| 2 | BN254_G2 | G2 of the BN254 curve |
| 3 | BLS12_381 | BLS 12-381 curve |
| 4 | BLS12_381_G1 | G1 of the BLS 12-381 curve |
| 5 | BLS12_381_G2 | G2 of the BLS 12-381 curve |

> A, B are curve points in G1 group. Each point consists of (X, Y) where X and Y are 256 bit integers, big-endian encoded. The encoded point is 64 bytes from concatenation of 32 byte X and 32 byte Y.

Reviewer: ```suggestion
A, B are curve points in affine representation: `X` concatenated to `Y`.

| Index | Name | Length of X or Y | Length of A or B |
| ----- | ------ | ---------------- | ----------------- |
| 1 | BN254_G1 | 32 bytes = 256 bits | 64 bytes = 512 bits |
| 2 | BN254_G2 | 64 bytes = 512 bits | 128 bytes = 1024 bits | 
| 4 | BLS12_381_G1 | 48 bytes = 384 bits | 96 bytes = 768 bits |
| 5 | BLS12_381_G2 | 96 bytes = 768 bits | 192 bytes = 1536 bits |

Checks:
* `ec_add` panics if `A` and `B` are not on the curve.
* `ec_add` does not verify that `A` and `B` are in the main prime-order subgroup.
```

Note for some reason https://eips.ethereum.org/EIPS/eip-2537 pads X/Y with 16 bytes for BLS12-381 G1 and with 32 bytes for G2. I don't think we need it.

Author: I think this is covered now.  It's annoying to make the table have different headings (they are generated). So the info is in the note field of the table.

## ec_scalar_mul v

- Opcode: 0xe1 {uint8 curve}
- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
- for (curve point A, scalar K) return the curve point KA
- **Cost**:  BN254=0 BN254_G1=100 BN254_G2=100 BLS12_381=0 BLS12_381_G1=200 BLS12_381_G2=200
- Availability: v9

> A is a curve point in G1 Group and encoded as described in `ec_add`. Scalar K is a big-endian encoded big integer that has no padding zeros.

Reviewer: Same checks as for `ec_add`.

Do you want to bound the maximum size of `K`?
Why restricting `K` to have no padding zeros?

Author: I don't see a reason for the padding thing. I removed it.

I'll have to benchmark this for large K. I'm guessing the cost is log(K)? If restricted, what is a reasonable restriction, K maybe can be 32 or 64 bytes? (does it just need to be as big as the order of the group?)

Author: Commenting before resolving: The limit for scalars is 32 bytes.  This might have to become curve specific someday, if we have curves larger than that.

## ec_pairing_check v

- Opcode: 0xe2 {uint8 curve}
- Stack: ..., A: []byte, B: []byte &rarr; ..., uint64
- for (points in G1 group G1s, points in G2 group G2s), return whether they are all paired => {0 or 1}
- **Cost**:  BN254=1000 BN254_G1=0 BN254_G2=0 BLS12_381=2000 BLS12_381_G1=0 BLS12_381_G2=0
- Availability: v9

> G1s are encoded by the concatenation of encoded G1 points, as described in `ec_add`. G2s are encoded by the concatenation of encoded G2 points. Each G2 is in form (XA0+i*XA1, YA0+i*YA1) and encoded by big-endian field element XA0, XA1, YA0 and YA1 in sequence.

Reviewer: ```suggestion
G1s are encoded by the concatenation of encoded G1 points, as described in `ec_add`. G2s are encoded by the concatenation of encoded G2 points. Each G2 is in form (XA0+i*XA1, YA0+i*YA1) and encoded by big-endian field element XA0, XA1, YA0 and YA1 in sequence.

Checks:
* `ec_pairing_check` panics if `A` and `B` do not contain the same number of points.
* `ec_pairing_check` panics if any point in `A` or `B` are not on there respective curve.
* `ec_pairing_check` panics if any point in `A` or `B` are not on there respective prime-order subgroup. (Note that this is a different behavior `ec_add` and `ec_scalar_mul`.
```

| 0 | BlkSeed | []byte |  |
| 1 | BlkTimestamp | uint64 |  |

## ec_add g

- Opcode: 0xe0 {uint8 curve}
- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
- for curve points A and B, return the curve point A + B
- **Cost**:  BN254g1=310 BN254g2=430 BLS12_381g1=540 BLS12_381g2=750
- Availability: v9

`EC` Groups:

| Index | Name | Notes |
| - | ------ | --------- |
| 0 | BN254g1 | G1 of the BN254 curve. Points encoded as 32 byte X following by 32 byte Y |
| 1 | BN254g2 | G2 of the BN254 curve. Points encoded as 64 byte X following by 64 byte Y |
| 2 | BLS12_381g1 | G1 of the BLS 12-381 curve. Points encoded as 48 byte X following by 48 byte Y |
| 3 | BLS12_381g2 | G2 of the BLS 12-381 curve. Points encoded as 96 byte X following by 48 byte Y |

> A and B are curve points in affine representation: X concatenated with Y. Fails if A or B is not in G. Does _not_ check if A and B are in the main prime-order subgroup.

Reviewer: ```suggestion
A and B are curve points in affine representation: field element X concatenated with field element Y. 

Encoding of a field element `Z` is as follows.
For base field elements (Fp), `Z` is represented as big-endian number and must be lower than the field modulus p.
For the quadratic field extension (Fp2), `Z` is encoded as the concatenation of the individual encoding of the coefficients. For an Fp2 element of the form `Z = Z0 + Z1 i`, where `i` is a formal quadratic non-residue, the encoding of Z is the concatenation of the encoding of `Z0` and `Z1` in this order. (In particular, it fails is `Z0` or `Z1` is larger or equal to p).

The point at infinity is encoded as `(X,Y) = (0,0)`.
Groups G1 and G2 are denoted addively.

Fails if A or B is not in G. 
(A and/or B are allowed to be the point at infinity.)
Does _not_ check if A and B are in the main prime-order subgroup.
```

taken from https://eips.ethereum.org/EIPS/eip-2537, needed to ensure canonicality.

## ec_scalar_mul g

- Opcode: 0xe1 {uint8 curve}
- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte

> - for curve point A and scalar B, return the curve point BA

Reviewer: ```suggestion
- for curve point A and scalar B, return the curve point BA, the point A multiplied by the scalar B.
```

- **Cost**:  BN254g1=2200 BN254g2=4460 BLS12_381g1=3640 BLS12_381g2=8530
- Availability: v9

A is a curve point encoded and checked as described in `ec_add`. Scalar B is interpreted as a big-endian unsigned integer. Fails if B exceeds 32 bytes.

## ec_pairing_check g

- Opcode: 0xe2 {uint8 curve}
- Stack: ..., A: []byte, B: []byte &rarr; ..., uint64

> - 1 if all points in A (concatenated points in group G) are paired with the corresponding points of B (concatenated points in the associated group), else 0

Reviewer: ```suggestion
- 1 if the product of the pairing of each point in A with its respective point in B is equal to the identity element of the target group Gt, else 0
```

Author: Thanks, I wrote that before I had any idea what pairing did!

- **Cost**:  BN254g1=18000 BN254g2=18000 BLS12_381g1=15000 BLS12_381g2=15000
- Availability: v9

A and B are concatenated points, encoded and checked as described in `ec_add`. A contains points of the group G, B contains points of the associated group (G2 if G is G1, and vice versa). Fails if A and B have a different number of points, or if any point is not in its described group or outside the main prime-order subgroup - a stronger condition than other opcodes.

## ec_multi_exp g

- Opcode: 0xe3 {uint8 curve}
- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
- for curve points A and scalars B, return curve point B0A0 + B1A1 + B2A2 + ... + BnAn
- **Cost**:  BN254g1=800 BN254g2=1800 BLS12_381g1=1400 BLS12_381g2=3500
- Availability: v9

A is a list of concatenated points, encoded and checked as described in `ec_add`. B is a list of concatenated 32 bytes scalars.

> 

Reviewer: ```suggestion
Contrary to `ec_scalar_mul`, all scalars must be exactly 32-bytes in `ec_multi_exp`.

Note: the name `ec_multi_exp` comes from https://eips.ethereum.org/EIPS/eip-2537 and GNARK, but a more consistent name would be `ec_multi_scalar_mul`.
```

Author: I'd rather avoid explicitly referencing the EIP and GNARK directly.  How about:
The name `ec_multi_exp` was chosen to reflect common usage, but a more consistent name would be `ec_multi_scalar_mul`?

Separately. Should we just get rid of `ec_scalar_mul`? You should get the exact same answer using this opcode with one point and one argument. (Internally, surprisingly, we might still call a different gnark routine for the singleton case because it's faster.). This also gets rid of the discussion of the different scalar sizes.

It we do, could we call it `ec_dot_product` and also avoid this discussion?

Reviewer 2: I up the proposal of renaming to `ec_dot_prod` than `ec_multi_exp`, and probably merge this one against `ec_scalar_mult`, for that is just one entry in vector multiplying another entry in another vec.

We might want to have linear cost model for this one? For the computation varies by the byte length.

Author: Currently proposing `ec_multi_scalar_mul` is lower conversation, resolving this one.

## ec_subgroup_check g

- Opcode: 0xe4 {uint8 curve}
- Stack: ..., A: []byte &rarr; ..., uint64
- whether point A is in the main prime-order subgroup of G => {0 or 1}. Program fails if A is not in G at all.
- **Cost**:  BN254g1=50 BN254g2=11500 BLS12_381g1=5600 BLS12_381g2=7100
- Availability: v9

> 

Reviewer: ```suggestion

Succeeds for the point at infinity.
```

(I did not check the code yet, but if there is no unit test for that corner case, we definitely need one.)

Author: Added, including unit test.
```
func TestSubgroupCheckInfinity(t *testing.T) {
	partitiontest.PartitionTest(t)
	t.Parallel()

	testAccepts(t, "int 64; bzero; ec_subgroup_check BN254g1;", pairingVersion)
	testAccepts(t, "int 128; bzero; ec_subgroup_check BN254g2;", pairingVersion)
	testAccepts(t, "int 96; bzero; ec_subgroup_check BLS12_381g1;", pairingVersion)
	testAccepts(t, "int 192; bzero; ec_subgroup_check BLS12_381g2;", pairingVersion)
}

```

## ec_map_to g

- Opcode: 0xe5 {uint8 curve}
- Stack: ..., A: []byte &rarr; ..., []byte
- maps field element A to group G
- **Cost**:  BN254g1=1700 BN254g2=11000 BLS12_381g1=5600 BLS12_381g2=43000
- Availability: v9

> BN254 points are mapped by the SVDW map. BLS12-381 points are mapped by the SSWU map. G1 element inputs are encoded single big-endian byte-array of length <= n.  G2 inputs are encoded as concatenated n-byte big-endian encoded integers. n == 32 for BN254 and 48 for BLS12-381. No input may exceed the modulus of the group. 

Reviewer: ```suggestion
BN254 points are mapped by the SVDW map. BLS12-381 points are mapped by the SSWU map. 

G1 element inputs are base field elements and G2 element inputs are quadratic field elements, with the same encoding rules as defined in `ec_add`.
In particular, no input coefficient can exceed the modulus of the group.
```

Author: `ec_add` takes G1 points, which are two concatenated 32 big-endian integers.  But here, we take the "base field elements" which for G1, are just single integers. So it does not seem helpful discuss the "same encoding rules as `ec_add`".

In fact, there is quite a departure from those rules, even when you consider the elements themselves. Here, at least for G1 groups, we allow short bytes sequences like 0x0108, and interpret is as 264 (256+8). We don't mind that the input is not 32 bytes long. Since `ec_add` effectively takes two such elements, it _demands_ the input is 64 bytes - two concatenated 32 byte values.

For `ec_map_to` applied to G2 fields, we can not allow such conveniences, since the base field is itself quadratic.  They must be 64 or 94 bytes (2 32s, or 2 48s).  That why I talk about "<= n" for G1 inputs, and exactly n for G2 inputs.

- Bytecode: 0xe2 {uint8}
- Stack: ..., A: []byte, B: []byte &rarr; ..., uint64
- 1 if the product of the pairing of each point in A with its respective point in B is equal to the identity element of the target group Gt, else 0
- **Cost**:  BN254g1=1 + 180 per 64 bytes of B; BN254g2=1 + 180 per 128 bytes of B; BLS12_381g1=450 + 400 per 96 bytes of B; BLS12_381g2=450 + 400 per 192 bytes of B
- Availability: v10

A and B are concatenated points, encoded and checked as described in `ec_add`. A contains points of the group G, B contains points of the associated group (G2 if G is G1, and vice versa). Fails if A and B have a different number of points, or if any point is not in its described group or outside the main prime-order subgroup - a stronger condition than other opcodes AVM values are limited to 4096 bytes, so `ec_pairing_check` is limited by the size of the points in the groups being operated upon.

## ec_multi_exp

- Syntax: `ec_multi_exp G` ∋ G: [EC](#field-group-ec)
- Bytecode: 0xe3 {uint8}
- Stack: ..., A: []byte, B: []byte &rarr; ..., []byte
- for curve points A and scalars B, return curve point B0A0 + B1A1 + B2A2 + ... + BnAn

> - **Cost**:  BN254g1=80 + 3 per 32 bytes of B; BN254g2=180 + 9 per 32 bytes of B; BLS12_381g1=140 + 4 per 32 bytes of B; BLS12_381g2=350 + 18 per 32 bytes of B

Author: Correct, this is vastly underpriced (I think 1000x less?) so I could do tests, but now that we have logicsig pooling, I should be able to revamp the tests using proper costs.

> - **Cost**:  BN254g1=80 + 3 per 32 bytes of B; BN254g2=180 + 9 per 32 bytes of B; BLS12_381g1=140 + 4 per 32 bytes of B; BLS12_381g2=350 + 18 per 32 bytes of B

Reviewer 2: got it, was thinking about introduction of logicsig pooling, so I should probably revisit the cost later.

Resolving this thread now.
